source/common/config/config_provider_impl.h

```diff
@@ -184,25 +190,48 @@ class ConfigSubscriptionCommonBase : protected Logger::Loggable<Logger::Id::conf
-   *         bound MutableConfigProviderCommonBase or nullptr when there are none.
-   */
-  MutableConfigProviderCommonBase* getAnyBoundMutableConfigProvider() const {
-    return !mutable_config_providers_.empty() ? *mutable_config_providers_.begin() : nullptr;
-  }
-
 protected:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    explicit ThreadLocalConfig(ConfigProvider::ConfigConstSharedPtr initial_config)
+        : config_(std::move(initial_config)) {}
+
+    ConfigProvider::ConfigConstSharedPtr config_;
+  };
+
   ConfigSubscriptionCommonBase(const std::string& name, const uint64_t manager_identifier,
                                ConfigProviderManagerImplBase& config_provider_manager,
-                               TimeSource& time_source, const SystemTime& last_updated,
-                               const LocalInfo::LocalInfo& local_info)
-      : name_(name),
+                               Server::Configuration::FactoryContext& factory_context)
+      : name_(name), tls_(factory_context.threadLocal().allocateSlot()),
         init_target_(fmt::format("ConfigSubscriptionCommonBase {}", name_), [this]() { start(); }),
         manager_identifier_(manager_identifier), config_provider_manager_(config_provider_manager),
-        time_source_(time_source), last_updated_(last_updated) {
-    Envoy::Config::Utility::checkLocalInfo(name, local_info);
+        time_source_(factory_context.timeSource()),
+        last_updated_(factory_context.timeSource().systemTime()) {
+    Envoy::Config::Utility::checkLocalInfo(name, factory_context.localInfo());
+  }
+
+  /**
+   * Propagates a config update to worker threads.
+   *
+   * @param update_fn the callback to run on each thread, it takes the previous version Config and
+   * returns a updated/new version Config.
+   * @param complete_cb the callback to run when the update propagation is done.
+   */
+  void applyConfigUpdate(
+      const std::function<
+          ConfigProvider::ConfigConstSharedPtr(ConfigProvider::ConfigConstSharedPtr)>& update_fn,
+      const Event::PostCb& complete_cb = []() {}) {
+    // It is safe to call shared_from_this here as this is in main thread, and destruction of a
+    // ConfigSubscriptionCommonBase owner (i.e., a provider) happens in main thread as well.
+    auto shared_this = shared_from_this();
+    tls_->runOnAllThreads(
+        [this, update_fn]() {
+          tls_->getTyped<ThreadLocalConfig>().config_ = update_fn(this->getConfig());
+        },
+        /*During the update propagation, a subscription may get teared down in main thread due to
+        all owners/providers destructed in a xDS update (e.g. LDS demolishes RouteConfigProvider and
+        its subscription). Hold a reference to the shared subscription instance to make sure the
+        update can be safely pushed to workers in such an event.*/
+        [shared_this, complete_cb]() { complete_cb(); });
```

Comment:
Shouldn't the capture be in the first lambda above, where `this` is being referenced, rather than here?

Reply:
I don't think so. If the race ever happens, we need to make sure "this" gets destructed on the main thread, as the subscription contains a tls_ instance. Capturing "this" in the complete_cb of runOnAllThreads makes sure that "*this" is alive after every first-lambda is called, and the held "*this" is posted back to main thread. 

If we capture it in the first lambda, the shared_this may get destructed in a worker thread, cause an assertion failure.

Reply:
Can you make this more explicit in the comment? This is a subtle correctness condition, but I think you are right.

Reply:
Sounds good. Comment updated!